src/tasks/components/NewTaskScheduler/ScriptSelector.tsx

```diff
@@ -0,0 +1,127 @@
+import React, {FC} from 'react'
+import {Link, useParams} from 'react-router-dom'
+import {
+  TechnoSpinner,
+  ComponentSize,
+  Dropdown,
+  EmptyState,
+  IconFont,
+  Input,
+  Form,
+} from '@influxdata/clockface'
+
+// Types
+import {RemoteDataState} from 'src/types'
+
+interface Props {
+  loading: RemoteDataState
+  scripts: any
+  selectedScript: any
+  searchTerm: string
+  setSelectedScript: (script: string) => void
+  handleSearchTerm: (str: string) => void
+}
+
+const ScriptSelector: FC<Props> = ({
+  loading,
+  scripts,
+  selectedScript,
+  setSelectedScript,
+  handleSearchTerm,
+  searchTerm,
+}) => {
+  const {orgID} = useParams<{orgID: string}>()
+
+  let list
+  if (
+    loading === RemoteDataState.NotStarted ||
+    loading === RemoteDataState.Loading
+  ) {
+    list = (
+      <div>
+        <TechnoSpinner strokeWidth={ComponentSize.Small} diameterPixels={32} />
+      </div>
+    )
+  }
+
+  if (loading === RemoteDataState.Error) {
+    list = (
+      <div>
+        <p>Could not get scripts</p>
+      </div>
+    )
+  }
+
+  if (loading === RemoteDataState.Done) {
+    list = (
+      <>
+        {scripts.map(script => (
+          <Dropdown.Item
+            key={script.id}
+            value={script.name}
+            onClick={() => setSelectedScript(script)}
+            selected={script.name === selectedScript?.name}
+          >
+            {script.name}
+          </Dropdown.Item>
+        ))}
+      </>
+    )
+    if (scripts.length === 0 && searchTerm) {
+      list = (
+        <EmptyState>
+          <p>{`No Scripts match "${searchTerm}"`}</p>
+        </EmptyState>
+      )
+    } else if (scripts.length === 0 && !searchTerm) {
+      list = (
+        <EmptyState>
+          <p>No Scripts found</p>
+        </EmptyState>
+      )
+    }
+  }
+  let buttonText = 'Select a Script'
+  if (loading === RemoteDataState.Done && selectedScript?.name) {
+    buttonText = selectedScript.name
+  }
+
+  return (
+    <div className="select-script">
+      <div className="create-task-titles script">Select a Script</div>
+      <Form.Element label="Script" required={true} className="script-dropdown">
+        <Dropdown
+          button={(active, onClick) => (
+            <Dropdown.Button
+              active={active}
+              onClick={onClick}
+              testID="variable-type-dropdown--button"
+            >
+              {buttonText}
+            </Dropdown.Button>
+          )}
+          menu={onCollapse => (
+            <Dropdown.Menu onCollapse={onCollapse}>
+              <Input
+                icon={IconFont.Search_New}
+                size={ComponentSize.Small}
+                value={searchTerm}
+                placeholder="Search Scripts"
+                onChange={evt => handleSearchTerm(evt.target.value)}
```

Comment:
I think we want to change how this system works. Right now, the task scheduler fetches the scripts and passes them down to the scripts selector, which couples the ScriptSelector to the task scheduler. Or more generically, it forces whatever component uses the ScriptSelector to fetch scripts and re-implement searching/filter on them. The way this event handler works currently, it'll re-fetch all the scripts every time there's a keystroke (after some debouncing time), after fetching all scripts up to `limit` on component load.

I think what this system should do instead is that this component should fetch all the scripts once, then when the user types a script name in the search box, this should filter the names of the scripts it already fetched. This will be much more efficient in network usage (it'll only make a single request, rather than up to `n` requests where `n` is number of keystrokes) and it should be much faster, since it'll filter an in-memory list, rather than having to make a trip across the newtork.

Take a look at this [blog post](http://highscalability.com/blog/2011/1/26/google-pro-tip-use-back-of-the-envelope-calculations-to-choo.html) to get an idea of the scale of network operations. If we assume this data isn't in the CPU cache but is in main memory, and that we're getting worst-performance when accessing it from memory, the differences in filtering in memory vs making a network request look like this:

```
Read 1 MB sequentially from memory 250,000 ns
Read 1 MB sequentially from network 10,000,000 ns
````

That's two orders of magnitude slower to fetch something across the network.